### ISSUE TITLE
Switch to "off the shelf" parser – yay or nay?!

### ISSUE THREAD

**Issue by eriksvedang** [coherence/protocol-code-generator#178]
Several members of the team has brought up the idea of switching our hand-written schema parser out for some kind of library that handles a stock file format, such as JSON or YAML. This is an interesting idea, and potentially quite worthwhile. I have some thoughts about it though.

By adding to this issue I hope we can get a better quality discussion, where we can clearly see the pros and cons of the various options. I'll start:

- The current parser is not super difficult to extend in certain directions – adding more keywords etc is quite clean. Its main problem is that it's very locked in to using significant whitespace (no braces) and a very rigid version of that, too (2 spaces only). Making it more flexible would complicate it a lot, so that's a big downside!

- To state the obvious – if we start depending on a library it has to be a great one, with a lot of users and ongoing maintenance. There are lots of parsers out there that people write and then abandon. For a switch to make sense we have to be pretty certain that we won't have to fork and maintain the library.

- This blog post explains pretty well why I'd advice against switching to JSON in particular: [Why JSON isn't a good configuration language](https://www.lucidchart.com/techblog/2018/07/16/why-json-isnt-a-good-configuration-language/)

- The current schema format is in my opinion quite elegant and professional looking. Switching to a *configuration file format* (which is what JSON/YAML is) will make the schema feel less like a programming language and create a bigger disconnect between defining the schema and then using the generated structs inside your program. This might be fine, but it's a usability concern. Any suggestion for another syntax should show how that syntax would actually look in the schema file, so that we can compare!

- In my opinion, a better syntax to consider is something that looks like C/C# – ideally that would allow the programmer to cut and paste an existing component struct from C# to the schema with minimal changes. That would be a great workflow win when converting existing ECS code to coherence. Here's an example of how that could look:

```c
component Health {
  int Value;
}
```

- One complaint with the current parser has been error messages. Various parsing libraries have very different quality to their error messages and the ability to configure those, so we must make sure that whatever we start using actually is better in that respect. When having a customised parser like the current one it's pretty simple to add specific error messages for our own use cases, this can become difficult with a more general solution.

**Comment by tadej** [comment]
I agree fully with your comments, I am against using JSON for this.

As a reference, please look at the SpatialOS schema: https://documentation.improbable.io/sdks-and-data/docs/schemalang-reference

It uses a domain-specific language (similar to what you propose above), not a data file format.

**Comment by cbrisebois** [comment]
So here are my thoughts

- the inflexible syntax is a huge red flag for me, anything that a user can do to break our syntax they will do so we want to reduce the amount of support we have to give / edge cases we have to fix because we're inventing a syntax from scratch. 

- we need to be careful that we're agreeing on what the schema does / is.  If we make it **look** like a programming language then we have to expect users to use it like a language.  I think this [example](https://documentation.improbable.io/sdks-and-data/docs/schemalang-reference#section-reusable-data-types) from Improbable's schemalang is a good example of what I'm talking about:
```
type SomeData {
  int32 value = 1;
}

component ThisComponentWontCompile {
  id = 1234;
  data SomeData;
  int32 extra_property = 2;
}

type OtherData {
  int32 value = 3;
}

component ThisComponentAlsoWontCompile {
  id = 1337;
  data SomeData;
  data OtherData;
}
```
This won't compile because it's mixing data type and property definitions but is super unclear because it almost looks like something you could do in C#. 

- If we want it to act like a programming language (type checking, methods) then we should just straight-up use C# or lua or some scripting language instead of trying to almost build our own.

- if we do want it to **look** like a language but with our own syntax then we should consider Lex / Yacc instead of our own parsing system since there's so much work in that field it's probably going to save us a ton as we expand our language

- If what we want is really a parsed data structure (what I'd argue we currently have rather than a domain-specific language) then we'd be much smarter to stick with something that is common and familiar even if it's less pretty (I don't think Scrawl is pretty at all).  I agree JSON is not the right tool, but YAML probably is.  Using something like YAML will immediately take work off our backs since there are linters and IDE supports for them and tons of documentation and examples which takes a huge support burden off of us.

```
name Schema

component Score
  number Int

component Health
  number Int

component ProjectileType
  number Int

component Velocity
  x Float
  y Float

command ApplyDamage
  change Int

command AwardPoints
  change Int
  
component PlayerName
  name String64
  
archetype Ship [priority "high"]
  lod 0
    WorldPosition
    WorldOrientation
    Velocity
    Health
    Score
    PlayerName
    
archetype Projectile [priority "high"]
  lod 0
    WorldPosition
    WorldOrientation
    ProjectileType
```

Example YAML: I used [YAML Lint](http://www.yamllint.com/) to help format it
```
Schema: name

Commands: 
  - 
    change: Int
    name: ApplyDamage
  - 
    change: Int
    name: AwardPoints

Components: 
  - 
    name: Score
    number: Int
  - 
    name: Health
    number: Int
  - 
    name: ProjectileType
    number: Int
  - 
    name: Velocity
    x: Float
    y: Float
  - 
    name: PlayerName
    value: String64

Archetypes: 
  - 
    lods: 
      0: 
        - WorldPosition
        - WorldOrientation
        - Velocity
        - Health
        - Score
        - PlayerName
    name: Ship
    priority: high
  - 
    lods: 
      0: 
        - WorldPosition
        - WorldOrientation
        - ProjectileType
    name: Projectile
    priority: high
```

**Comment by tadej** [comment]
One thing to keep in mind here is that we **do** intend the schema to contain **code** at some point.

When running an MMO, the schema will sometimes change, and this means **schema versioning** can suddenly be very useful.

Why? Because the game client might be large and a pain to update so the developer may want some backwards compatibility between different schemas.

E.g. (single replicator for simplicity):

1. The Replicator is running schema v1.2.
2. Simulator01 is still on schema v1.1. 
3. Simulator02 is on schema v1.2.
4. Client01 is on schema v1.2
5. Client02 is on schema v1.0.

So schemas have evolved from v1.0 -> v1.1 -> v1.2 and different clients support different version.

Let's take for example a component called _FacialExpression_. 

- v1.0 had none (graphics didn't support them yet), 
- v1.1 had 5 possible facial expressions, 
- v1.2 now has 20 possible expressions (or integers running between [0,19]).

Scenarios:

1. Client02 sees an entity simulated on Simulator01 or Simulator02: it ignores _FacialExpression_.
2. Client01 sees an entity simulated on older version of the schema: we need some mapping between no _FacialExpression_ or _FacialExpression_[0,4] to _FacialExpression_[0,19].

What this means is that the schema will have to support things like **conditionals** (if statements, switches, etc.) as well.

In theory, all of this is possible with YAML and Scrawl, but I'm afraid it might end up looking really messy really fast.

**Comment by ottov1** [comment]
If we support versions like described above then I think things will get messy really fast no matter what format we choose. The questions is - do we really need to? All the online games I've played, especially MMO games, force their clients to the latest versions when there's an update. Also, Improbable doesn't seem to support versioning of the schema file either: https://documentation.improbable.io/sdks-and-data/lang-re/docs/backward-compatibility.

Speaking of the format, we can have more than one format. One is facing the user - it is the format that is easy to edit by a human. The other formats are the compiled forms of the schema that can be fed to the client, simulator and replicator, much like what the protocol generator is doing. For example, if we end up with something trully complex, then we could compile the schema to a JSON/YAML file and feed it to the replication server instead of adding the (presumably complex) schema parser there too.

**Comment by tadej** [comment]
> If we support versions like described above then I think things will get messy really fast no matter what format we choose.

Not necessarily. The idea is **not** to maintain backwards compatibility forever, but only a few versions back.
One benefit is also to not have to shut the whole system down when upgrading, and to allow it to happen gracefully and gradually. It's not just the clients, but all the simulators and replicators as well.

Granted, this could also be achieved in a separate file with a scripting language, while keeping something like YAML/Scrawl for the schema itself.

This feature is not high on our priority list but it has been in our plans for a while. Happy to discuss.

> Also, Improbable doesn't seem to support versioning of the schema file either: https://documentation.improbable.io/sdks-and-data/lang-re/docs/backward-compatibility.

Which was one of the major complaints we got from their (ex-)customers. 

To clarify, Improbable is a cautionary tale and they are doing a lot of things wrong. 
We are **not** trying to copy them, we are doing our own thing. 

However, they have been dealing with customers for a few years and changed their entire SDK a few times.
Things like how they handle types and names in their schema can sometimes be informative.

What they have **not** been doing is supporting many running games, with notable exceptions like Worlds Adrift by Bossa (a developer we're speaking to regularly), that was shut down.

**Comment by ottov1** [comment]
> One benefit is also to not have to shut the whole system down when upgrading, and to allow it to happen gracefully and gradually. 

There are other ways to achieve this without using versioning. Happy to discuss this as well when we get there :)

Back to the schema format - if we do decide to go with a declarative format like YAML then I'd suggest using TOML instead. It is quite popular, has wide support, and is easy to parse: https://toml.io/en/.

**Comment by eriksvedang** [comment]
> the inflexible syntax is a huge red flag for me, anything that a user can do to break our syntax they will do so we want to reduce the amount of support we have to give / edge cases we have to fix because we're inventing a syntax from scratch.

Yes, I agree – I'm not overly sold on whitespace sensitive syntax myself, especially since you can't auto format it.

>  If we make it look like a programming language then we have to expect users to use it like a language

Maybe, but that should be solvable with documentation and helpful error messages that anticipates common errors. Initializing components like that doesn't work with Unity ECS anyway, so it's not something people would expect I hope.

> If we want it to act like a programming language (type checking, methods) then we should just straight-up use C# or lua or some scripting language instead of trying to almost build our own.

Agreed.

> if we do want it to look like a language but with our own syntax then we should consider Lex / Yacc instead 

Yeah, if there's a good Go alternative that's definitely a viable option – having the grammar defined in a declarative way is nice. My anecdotal knowledge is that a lot of projects that start with lex/yacc eventually end up with a hand written parser, very much like the one we have – I think due to better error messages (and sometimes speed). Parsing text is one of the easier parts of a language implementation after all, and lex/yacc is pretty arcane by todays standard. Perhaps the Go alternatives are better though!

> Syntax example:

```
  - 
    name: Velocity
    x: Float
    y: Float
```

This looks fine, but it uses the keys of the dictionary as both keywords (`name`) and as the names for the fields (`x`, `y`). That won't work if you want a field called `name` for example, right? Also, I wonder how to add meta data in a clean way using that approach. A less ambiguous structure would be:

```
  - 
    name: Velocity
    fields: 
      - name: x
      - type: Float
      - meta: 
          - resolution: 4
... and so on...
```

But that's a lot more annoying to write, of course. Maybe there's another way that is better, or TOML could handle it better..?

> One thing to keep in mind here is that we do intend the schema to contain code at some point.

I did not know that. I *really* don't think we should implement a complete language (even though it's a lot of fun :D).

> Speaking of the format, we can have more than one format. 

Yes, this approach could be good since we could experiment easily with new features by only adding them to the data file, and it makes the user-facing language easier to implement – it's just a text transformation. It complicates the toolchain a bit though.
